Update release_notes.rst (#1270)
cocotb/cocotb: documentation/source/release_notes.rst

All releases are available from the `GitHub Releases Page <https://github.com/cocotb/cocotb/releases>`_.

cocotb 1.3
==========

Released on 31 December 2019

New features
------------
* :func:`cocotb.triggers.with_timeout`, for a shorthand for waiting for a trigger with a timeout.
* New ``timeout_time`` and ``timeout_unit`` arguments to :func:`cocotb.test`, for adding test timeouts.
* The ``expect_error`` argument to :func:`cocotb.test` now accepts a specific exception type.

> * :envvar:`COCOTB_RESULTS_FILE`, to allow configration of the xunit XML output filename.

Reviewer: A ``make changes`` in ``documentation/`` shows that ``COCOTB_HDL_TIMEUNIT`` and ``COCOTB_HDL_TIMEPRECISION`` have also been added, and says that ``raise_error`` and ``create_error`` are deprecated in favor of ``TestError``.
Also, Verilator support!

Author: Mind adding the bullet points you'd like to see added in a comment?

Reviewer: Maybe like this:
```
* New makefile variables :make:var:`COCOTB_HDL_TIMEUNIT`/:make:var:`COCOTB_HDL_TIMEPRECISION` for setting the default time unit/precision that should be assumed for simulation when not specified by modules in the design.
```
```
* Support for the :ref:`Verilator` simulator (in version 4.020 and above.)
```

Author: Probably worth a new deprecations subheading

All releases are available from the `GitHub Releases Page <https://github.com/cocotb/cocotb/releases>`_.

cocotb 1.3
==========

Released on 1 January 2020

New features
------------
- :func:`cocotb.triggers.with_timeout`, for a shorthand for waiting for a trigger with a timeout. (:pr:`1119`)
- New ``timeout_time`` and ``timeout_unit`` arguments to :func:`cocotb.test`, for adding test timeouts. (:pr:`1119`)
- The ``expect_error`` argument to :func:`cocotb.test` now accepts a specific exception type. (:pr:`1116`)

> - :envvar:`COCOTB_RESULTS_FILE`, to allow configration of the xunit XML output filename.  (:pr:`1053`)

Reviewer: Did you intentionally leave out the ``COCOTB_HDL_TIMEUNIT`` and ``COCOTB_HDL_TIMEPRECISION`` makevars?

Author: No, it just wasn't labelled appropriately so I couldn't find the PR before my train arrived.

Author: You can use a GitHub suggestion to add this bullet here, if you like

Reviewer: This is actually the first item in https://github.com/cocotb/cocotb/pull/1270#discussion_r362252238, or do you mean something else?

Author: I mean use the github suggestion button so that I can apply it from mobile.

> - :envvar:`COCOTB_RESULTS_FILE`, to allow configration of the xunit XML output filename.  (:pr:`1053`)

Reviewer: ```suggestion
- New environment variable :envvar:`COCOTB_RESULTS_FILE`, to allow configuration of the xunit XML output filename.  (:pr:`1053`)
- New makefile variables :make:var:`COCOTB_HDL_TIMEUNIT`/:make:var:`COCOTB_HDL_TIMEPRECISION` for setting the default time unit/precision that should be assumed for simulation when not specified by modules in the design. (:pr:`1113`)
```

> - :class:`cocotb.triggers.Timer` can now be used with :class:`fraction.Fraction` instances, allowing constructs like ``Timer(Fraction(1, 10000), units='sec')`` as an alternate spelling for ``Timer(100, units='us')``. (:pr:`1114`)

Reviewer: Should this be
```suggestion
- :class:`cocotb.triggers.Timer` can now be used with :class:`fractions.Fraction` instances, allowing constructs like ``Timer(Fraction(1, 10000), units='sec')`` as an alternate spelling for ``Timer(100, units='us')``. (:pr:`1114`)
```
? (https://docs.python.org/3/library/fractions.html#fractions.Fraction)

> - :class:`cocotb.triggers.Timer` can now be used with :class:`fractions.Fraction` instances, allowing constructs like ``Timer(Fraction(1, 10000), units='sec')`` as an alternate spelling for ``Timer(100, units='us')``. (:pr:`1114`)

Reviewer: ```suggestion
- :class:`cocotb.triggers.Timer` can now be used with :class:`fractions.Fraction` and :class:`decimal.Decimal` instances, allowing constructs like ``Timer(Fraction(1, 10000), units='sec')`` or ``Timer(Decimal('1e-9'), units='sec')`` as an alternate spelling for ``Timer(100, units='us')``. (:pr:`1114`)
```
Of course, feel free to disregard if you think this is too excessive.

Author: Did I actually fix it for decimal?

Reviewer: You have a test for it at least:
https://github.com/cocotb/cocotb/blob/47243e4f342cb95a9e7bcb7307978662a85eec14/tests/test_cases/test_cocotb/test_cocotb.py#L328-L331

- Support for the :ref:`Verilator` simulator (version 4.020 and above).
- A new ``bus_separator`` argument to :class:`cocotb.drivers.BusDriver`. (:pr:`1160`)

> - A new ``start_high`` argument to :class:`cocotb.clock.Clock`. (:pr:`1036`)

Reviewer: ```suggestion
- A new ``start_high`` argument to :meth:`cocotb.clock.Clock.start`. (:pr:`1036`)
```

> - A new ``start_high`` argument to :class:`cocotb.clock.Clock`. (:pr:`1036`)

Reviewer: I'm wondering whether even such small changes should get``versionadded`` in the docstring?

Author: In my opinion yes, they should.

Reviewer: Ok, part of #1276.

Notable changes and bug fixes
-----------------------------
- ``DeprecationWarning``\ s are now shown in the output by default.
- Tracebacks are now preserved correctly for exceptions in Python 2. The tracebacks in all python versions are now a little shorter.
- :func:`cocotb.external` and :func:`cocotb.function` now work more reliably and with fewer race conditions.

> - A failing ``assert`` will be considered a test failure. Previously, it was considered a test _error_.

Reviewer: ```suggestion
- A failing ``assert`` will be considered a test failure. Previously, it was considered a test *error*.
```

> - :meth:`~cocotb.handles.NonConstantObject.drivers` and :meth:`~cocotb.handles.NonConstantObject.drivers` now work correctly in python 3.7 onwards.

Reviewer: This is pointing two times to the same method (and doesn't hyperlink).

Author: `loads`, whoops

> - :meth:`~cocotb.handles.NonConstantObject.drivers` and :meth:`~cocotb.handles.NonConstantObject.drivers` now work correctly in python 3.7 onwards.

Reviewer: ```suggestion
- :meth:`~cocotb.handles.NonConstantObject.drivers` and :meth:`~cocotb.handles.NonConstantObject.loads` now work correctly in python 3.7 onwards.
```

Deprecations
------------

> - ``cocotb.result.raise_error`` and ``cocotb.result.create_error`` are deprecated in favor of using python exceptions directly.

Reviewer: ```suggestion
- ``cocotb.result.raise_error`` and ``cocotb.result.create_error`` are deprecated in favor of using Python exceptions directly.
```

- New makefile variables :make:var:`COCOTB_HDL_TIMEUNIT` and :make:var:`COCOTB_HDL_TIMEPRECISION` for setting the default time unit and precision that should be assumed for simulation when not specified by modules in the design. (:pr:`1113`)
- Support for the :ref:`Verilator` simulator (version 4.020 and above).
- A new ``bus_separator`` argument to :class:`cocotb.drivers.BusDriver`. (:pr:`1160`)
- A new ``start_high`` argument to :meth:`cocotb.clock.Clock.start`. (:pr:`1036`)
- A new :data:`cocotb.__version__` constant, which contains the version number of the running cocotb. (:pr:`1196`)

Notable changes and bug fixes
-----------------------------
- ``DeprecationWarning``\ s are now shown in the output by default.
- Tracebacks are now preserved correctly for exceptions in Python 2. The tracebacks in all python versions are now a little shorter.
- :func:`cocotb.external` and :func:`cocotb.function` now work more reliably and with fewer race conditions.
- A failing ``assert`` will be considered a test failure. Previously, it was considered a test *error*.

> - :meth:`~cocotb.handle.NonConstantObject.drivers` and :meth:`~cocotb.handle.NonConstantObject.loads` now work correctly in Python 3.7 onwards.

Reviewer: Does this mean "now also work correctly in Python 3.7 onwards" or "now work correctly in Python 3.7 onwards (but are incorrect in lower versions)."?

Author: No, simply that they _only_ worked in lower version before.

> - :meth:`~cocotb.handle.NonConstantObject.drivers` and :meth:`~cocotb.handle.NonConstantObject.loads` now work correctly in Python 3.7 onwards.

Reviewer: ```suggestion
- :meth:`~cocotb.handle.NonConstantObject.drivers` and :meth:`~cocotb.handle.NonConstantObject.loads` now also work correctly in Python 3.7 onwards.
```
